File: typescript/8.字符串转换整数-atoi.js
/*
 * @lc app=leetcode.cn id=8 lang=javascript
 *
 * [8] 字符串转换整数 (atoi)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
    let i = 0
    //丢弃前导空格
    while (s[i] == ' ') i++
    //判断是否是数字
    if (isNaN(s[i])) {
        if (s[i] != '-' && s[i] != '+') return 0
        else if (isNaN(s[i + 1])) return 0
    }

    let result = 0
    let MIN = -2147483648, MAX = 2147483647

    //判断正负号
    if (s[i] == '-') {
        result = - parseInt(s[++i]), i++
        while (i < s.length && !isNaN(s[i]) && s[i] != ' ')
            result = result * 10 - parseInt(s[i++])
    }
    else {
        if (s[i] == '+') i++
        while (i < s.length && !isNaN(s[i]) && s[i] != ' ')
            result = result * 10 + parseInt(s[i++])
    }

    if (result < MIN)
        result = MIN
    else if (result > MAX)
        result = MAX

    return result
}
// @lc code=end

File: c++/leetcode/8.字符串转换整数-atoi.cpp
/*
 * @lc app=leetcode.cn id=8 lang=cpp
 *
 * [8] 字符串转换整数 (atoi)
 */
#include <string>
#include <unordered_map>
using namespace std;
// @lc code=start

class Solution {
public:
    class Automation {
        enum STATE {
            START, SIGNED, NUMBER, END
        };
        unordered_map<STATE, vector<STATE>>table = {
            {STATE::START,{STATE::START,STATE::SIGNED,STATE::NUMBER,STATE::END}},
            {STATE::SIGNED,{STATE::END,STATE::END,STATE::NUMBER,STATE::END}},
            {STATE::NUMBER,{STATE::END,STATE::END,STATE::NUMBER,STATE::END}},
            {STATE::END,{STATE::END,STATE::END,STATE::END,STATE::END}}
        };
        STATE state = STATE::START;
        int col(char c) {
            if (c == ' ')return 0;
            else if (c == '+' || c == '-')return 1;
            else if (isdigit(c))return 2;
            return 3;
        }
    public:
        int sign = 1;
        long long ans = 0;
        void input(char c) {
            state = table[state][col(c)];
            if (state == STATE::NUMBER) {
                ans = ans * 10 + c - '0';
                ans = sign == 1 ? min(ans, (long long)INT_MAX) : min(ans, -(long long)INT_MIN);
            }
            else if (state == STATE::SIGNED)
                sign = c == '+' ? 1 : -1;
        }
    };
    int myAtoi(string s) {
        Automation automation;
        for (auto c : s)automation.input(c);
        return automation.sign * automation.ans;
    }
};
// @lc code=end

